drivers:lpadc:enable nxp lpadc acquisition time feature

ZhaoxiangJin opened the pull request:
Enable nxp lpadc acquisition time feature, acquisition time can be expressed in microseconds, nanosecond, and ADC tickets, this patch has been processed these three different situations. When the input acquisition time unit is ADC tickets, it only needs to be matched with the elements in the sample time array. If the input acquisition time unit is nanoseconds or microseconds, first convert the unit of the elements in the sample time array from ADCK to nanoseconds, and then use the dichotomy to match the option closest to the acquisition time from the sample time array.

decsny commented:
please read https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#contributor-expectations

this commit needs to be split up to multiple before it will be reviewed, see other PRs for examples

ZhaoxiangJin commented:
> please read https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#contributor-expectations
> 
> this commit needs to be split up to multiple before it will be reviewed, see other PRs for examples

ZhaoxiangJin commented:
> please read https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#contributor-expectations
> 
> this commit needs to be split up to multiple before it will be reviewed, see other PRs for examples

Thanks for your reminder, I updated the PR and split it into four commits, but the PR was accidentally closed by me. Can you please help me reopen it? Thanks.

ZhaoxiangJin committed:
samples: drivers: adc: Update acquisition-time property configuration

Enabled nxp lpadc driver acquisition-time feature, the overlay file
can express acquisition-time in microseconds, nanosecond and ADC
tickets.

The purpose of updating these overlay files is not only for local
testing, but also to remind users that they can use this feature.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
drivers: lpadc: enable nxp lpadc acquisition time feature

Enable nxp lpadc acquisition time feature, acquisition time can be
expressed in microseconds, nanosecond, and ADC tickets, this patch
has been processed these three different situations. When the input
acquisition time unit is ADC tickets, it only needs to be matched
with the elements in the sample time array. If the input acquisition
time unit is nanoseconds or microseconds, first convert the unit of
the elements in the sample time array from ADCK to nanoseconds, and
then use the dichotomy to match the option closest to the acquisition
time from the sample time array.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
dts: nxp: Add LPADC clocks properities to SoC dtsi

Add LPADC clock properties to SoC dtsi on adc/lpadc node

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
drivers: clock_control: add supoort for LPADC functional clock frequency

obtain on RTxxx and LPC platforms.

The lpadc driver needs to obtain its functional clock to configure
the acquisition-time. This patch add support for I.MX RT three digit
and LPC platforms.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented:
could anyone help to review this PR? Thanks a lot.

decsny commented on samples/drivers/adc/boards/lpcxpresso55s36.overlay:
```diff
@@ -26,7 +26,7 @@
 		zephyr,gain = "ADC_GAIN_1";
 		zephyr,reference = "ADC_REF_EXTERNAL0";
-		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
+		zephyr,acquisition-time = <ADC_ACQ_TIME(ADC_ACQ_TIME_TICKS, 7)>;
```
is there a reason for 7 in particular?

ZhaoxiangJin replied:
For the ADC, the longer the sampling time is set, the higher the accuracy of the data obtained, but too high a sampling time will cause the sampling rate to decrease, so I chose a more compromised configuration option.

decsny commented on dts/arm/nxp/nxp_lpc55S3x_common.dtsi:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2022 NXP
+ * Copyright 2022,2024 NXP
```
need space after comma

ZhaoxiangJin replied:
thanks for your reminder, I will update it.

decsny commented on dts/arm/nxp/nxp_rt6xx_common.dtsi:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, NXP
+ * Copyright (c) 2020,2024 NXP
```
space after comma and remove the (c)

ZhaoxiangJin replied:
thanks for your reminder, I will update it.

decsny commented on drivers/clock_control/clock_control_mcux_ccm.c:
```diff
@@ -377,6 +377,15 @@ static int mcux_ccm_get_subsys_rate(const struct device *dev,
+		*rate = CLOCK_GetRootClockFreq(kCLOCK_Root_Adc1);
+		break;
+	case IMX_CCM_LPADC2_CLK:
+		*rate = CLOCK_GetRootClockFreq(kCLOCK_Root_Adc2);
```
do all platforms have these two enums defined ? ie what if a platform has only 1 adc, will it cause a build failure, or do we have no cases like that yet (if not, then it's fine for now)

ZhaoxiangJin replied:
Currently, for all platforms equipped with LPADC and CCM controllers(RT116x, RT117x), the number of ADC instances is 2.

decsny commented on drivers/adc/adc_mcux_lpadc.c:
```diff
@@ -93,6 +190,12 @@ static int mcux_lpadc_channel_setup(const struct device *dev,
 
+	/* Configure LPADC acquisition time. */
+	if (mcux_lpadc_acquisition_time_setup(dev, channel_cfg->acquisition_time, cmd)) {
+		LOG_ERR(" LPADC acquisition time setting failed");
```
extra space in the beginning not needed

ZhaoxiangJin replied:
thanks for your reminder, I will remove the space.

decsny commented on drivers/adc/adc_mcux_lpadc.c:
```diff
@@ -57,7 +59,107 @@ struct mcux_lpadc_data {
+	uint32_t adc_freq_hz = 0;
+	float adc_cycle_ns = 0.0;
+	/* The unit of elements in the sample time array is ADCK. */
+	float sample_time_array[] = {3.0, 5.0, 7.0, 11.0, 19.0, 35.0, 67.0, 131.0};
```
why is it needed to use floating points here ?

ZhaoxiangJin replied:
In this patch logic, I convert the sample time array elements unit from ADCK to ns, that is each element needs to be multiplied by the variable  **adc_cycle_ns**, and the variable adc_cycle_ns is float point type.

decsny reviewed (dismissed):
The commits are well structured except for the fact that they are in a backwards order. You need to put the clock control commit first, then DTS changes, then the driver changes, then the sample overlays. This I think will ensure bisectibility of the commit history.

I still have to review the mcux_lpadc_acquisition_time_setup function more closely but first help me to understand why you are using floating point arithmetic so I have some better background about the code. Since floating point arithmetic will incur a cost on the platforms without an FPU especially there must be a reason?

decsny commented:
@yvanderv please review the copyright changes of files that did not have NXP copyright, I don't know the rule here

ZhaoxiangJin commented:
> The commits are well structured except for the fact that they are in a backwards order. You need to put the clock control commit first, then DTS changes, then the driver changes, then the sample overlays. This I think will ensure bisectibility of the commit history.
> 
> I still have to review the mcux_lpadc_acquisition_time_setup function more closely but first help me to understand why you are using floating point arithmetic so I have some better background about the code. Since floating point arithmetic will incur a cost on the platforms without an FPU especially there must be a reason?

Sorry for the late commit sequence update. The reason for using the float point type is the float point value are introduced when calculating the ADC period based on the ADC clock.

decsny commented:
I am still confused why does adc_cycle_ns even need to be a float? It is being used to calculate the float sample_time_array values, but then these values are later casted to an int when used. Can't all of these be long ints perhaps? Or am I missing something

ZhaoxiangJin commented:
> I am still confused why does adc_cycle_ns even need to be a float? It is being used to calculate the float sample_time_array values, but then these values are later casted to an int when used. Can't all of these be integer types perhaps? Or am I missing something

Why I think the variable **adc_cycle_ns** needs to be set to a floating point type, I would like to give an example to illustrate. Under NXP LPC55S36 platform, the default ADC clock frequency is 48MHz, so the ADC cycle is approximately 20.833 ns (if variable **adc_cycle_ns** is set to an integer type, then this number is 20). 

case1(float point type, 20.833):
sample_time_array[0] = 20.833 * 3 =  62.499 ns
sample_time_array[1] = 20.833 * 5 = 104.165 ns
if the parameter ADC acquisition time passed by the application layer is 82ns, then the option **kLPADC_SampleTimeADCK3** will be set to the corresponding bitfield.

case1(integer type, 20):
sample_time_array[0] = 20 * 3 =  60 ns
sample_time_array[1] = 20 * 5 = 100 ns
if the parameter ADC acquisition time passed by the application layer is 82ns, then the option **kLPADC_SampleTimeADCK5** will be set to the corresponding bitfield. 

If we use integer type in the code, the accuracy of the calculation results decreases. At this clock frequency (48MHz), and application configuration options (82ns), the optimal configuration item should be  **kLPADC_SampleTimeADCK3**.

Currently, my local code has changed the ADC sample time array (**sample_time_array**) to the integer type. The reason is that converting the final calculation result into an integer will not cause too much loss of accuracy, and at the same time reduces the calculation pressure of chips without FPU platforms (the subsequent calculation process uses array elements to compare with the **acquisition_time_value**).

for example
sample_time_array[0] = 20.833 * 3 =  62.499 ns = 62
sample_time_array[1] = 20.833 * 5 = 104.165 ns = 104
if the parameter ADC acquisition time passed by the application layer is 82ns, then the option **kLPADC_SampleTimeADCK3** will be set to the corresponding bitfield.

ZhaoxiangJin commented:
Recreate a new PR (https://github.com/zephyrproject-rtos/zephyr/pull/70636) based on the nxp-upstream repository.
